debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/ParallelSnapshotReader.java

```
@@ -0,0 +1,199 @@
+
+    }
+
+    // for testing purposes
+    /*package private*/ ParallelSnapshotReader(BinlogReader oldTablesBinlogReader,
+                                               SnapshotReader newTablesSnapshotReader,
+                                               BinlogReader newTablesBinlogReader) {
+        this.oldTablesReader = oldTablesBinlogReader;
+        this.newTablesBinlogReader = newTablesBinlogReader;
+        this.newTablesReader = new ChainedReader.Builder().addReader(newTablesSnapshotReader).addReader(newTablesBinlogReader).build();
+    }
+
+    /**
+     * Create and return a {@link ReconcilingBinlogReader} for the two binlog readers contained in this
+     * ParallelSnapshotReader.
+     * @return a {@link ReconcilingBinlogReader}
+     */
+    public ReconcilingBinlogReader createReconcilingBinlogReader() {
+        return new ReconcilingBinlogReader(oldTablesReader, newTablesBinlogReader);
+    }
+
+    @Override
+    public void uponCompletion(Runnable handler) {
+        uponCompletion.set(handler);
+    }
+
+    @Override
+    public void initialize() {
+        oldTablesReader.initialize();
+        newTablesReader.initialize();
+    }
+
+    @Override
+    public void start() {
+        if (running.compareAndSet(false, true)) {
+            oldTablesReader.start();
+            newTablesReader.start();
+        }
+    }
+
+    @Override
+    public void stop() {
+        boolean error = false;
+        try {
+            logger.info("Stopping the {} reader", oldTablesReader.name());
+            oldTablesReader.stop();
+        } catch (Throwable t) {
+            logger.error("Unexpected error stopping the {} reader", oldTablesReader.name());
+            error = true;
+        }
+
+        try {
+            logger.info("Stopping the {} reader", newTablesReader.name());
+            newTablesReader.stop();
+        } catch (Throwable t) {
+            logger.error("Unexpected error stopping the {} reader", newTablesReader.name());
```

Why is it that these exeptions are swallowed? And why shouldn't go the reader go to "stopped" state in case of an error?

---

My method here is just based on what's already being done in the ChainedReader. The exceptions are swallowed in the chainedReader as well. I did miss that in ChainedReader, stop is always set to be false, even if there is an error, so I'll fix that here.